### ISSUE TITLE
Can we put the weather data server URL into an easier configurable place?

### ISSUE THREAD

**Issue by billyswong** [mate-desktop/libmateweather#122]
The weather data fails to be retrieved again because of the remote server change. https://github.com/mate-desktop/libmateweather/pull/79#issuecomment-1765810415

If the URL isn't hardcoded in libmateweather.so, it will be a lot easier to get things corrected. We don't need to squeeze the last drop of efficiency for periodic weather information update.

**Comment by pepa65** [comment]
I guess this is the time to do it! Hopefully it will be easier for users to change the URL in case changes happen.

**Comment by billyswong** [comment]
Imagine the next time the new URL being longer than the old URL. Ad-hoc patching like what people are doing now won't work.

**Comment by raveit65** [comment]
New releases for 1.26 and master branch https://github.com/mate-desktop/libmateweather/releases

**Comment by billyswong** [comment]
Ad-hoc patching seems not working for some people running older Linux distributions.

For example, the 3rd party URL information can be put into an extra text-format file in `/usr/share/libmateweather`, where currently only contains the location data in XML.

**Comment by pepa65** [comment]
Excellent idea to use a currently already used file location for storing the URL(s) in!

**Comment by rob5n** [comment]
How is the average user going to know what the new weather service URL is?

What if the latest URL was stored in a file on github, and the app would occasionally pull the file down and store locally.  If it can't find the file on github then use the URL in the local file.

**Comment by pepa65** [comment]
That would be a great solution! It could be locally edited, but it could all happen automatically too..!

**Comment by highflux7** [comment]
Did the URL change yet again?!?   Data is not being retrieved as of 30-Oct-2023.

**Comment by lukefromdc** [comment]
With libmateweather from current git master I had no trouble showing weather for a nearby airport in the US

**Comment by procule** [comment]
> Ad-hoc patching seems not working for some people running older Linux distributions. If they don't get the patch from distribution repository at the end, they probably need to compile the library by themselves.

It was my case and even recompiling the library didn't work. My guess is that it is related to another linked library (libsoup) doing the request. I have sent a PR to resolve the issue: https://github.com/mate-desktop/libmateweather/pull/131

**Comment by pepa65** [comment]
I have patched my patch-hack to not include `www.` in the link, avoiding a 308 redirect which might be causing problems:
[patchmatelibweather](https://good4.eu/plw)

**Comment by bttrx** [comment]
> I have patched my patch-hack to not include `www.` in the link, avoiding a 308 redirect which might be causing problems: [patchmatelibweather](https://good4.eu/plw)

Your patch is wrong. The search pattern should include `www.`, but the replace pattern not. See https://forums.linuxmint.com/viewtopic.php?p=2399362#p2399362

**Comment by pepa65** [comment]
> Your patch is wrong. The search pattern should include `www.`, but the replace pattern not.

Absolutely right, thanks, good catch. It seems I patched my patch way too hastily.
Corrected now.

**Comment by L-U-T-i** [comment]
> How is the average user going to know what the new weather service URL is?
> 
> What if the latest URL was stored in a file on github, and the app would occasionally pull the file down and store locally. If it can't find the file on github then use the URL in the local file.

If the average user can not find a new weather service URL (by a google search for instance, or in this particular case through a server response code / redirect), he / she should:
a) get some professional help
b) wait "things to resolve by themselves" (by a package update, if he/she ia able at least to do that...)
I mean generally, for any computer issues.

Probably there wouldn't be /etc directory and/or dconf settings at all if such users would be a reference.

I think some simple way to change this URL (or any other setting that can change over night) would be a great addition, as one has a much more simple way to resolve even temporary changes. For such an average user, nothing would become more complicated though. Meaning a benefit for those who can use it, and no loss for the others.

**Comment by L-U-T-i** [comment]
It became obvious these days there should be at least some config file for libmateweather, probably in `/etc/mate` or `/etc/libmateweather` or so - including not only URL but probably also some other data to help resolve data retrieved from this URL (API related - like some regex sting to parse the data retrieved...).

This way, it would become much more robust, meaning no need to wait for a new release (what will become more problematic with time for 1.26 branch for instance). If one can find info about URL / API change (what should probably be published at aviationweather site, it seems developers there are very reasonable and cooperative, so I guess this shouldn't be a big problem to get...), it should be relatively simple for an average user to just update the configuration file and everything should work again.

**Comment by vkareh** [comment]
The problem with making the URL configurable is that it assumes that a change in URL does not come with a change in payload schema.

This last change required changes in how we handle the response we get back from NOAA, and just changing the URL would not have fixed the problem.

I think a more resilient approach is to support fetching weather data from different sources. Those sources can be hard-coded, but if one fails, you can use the next one.

**Comment by L-U-T-i** [comment]
I agree more sources would make it more resilient. It may be however better if user can set the preference(s) - maybe some sources get more frequent updates for a particular location, or provide more accurate data or so.

But I still believe it should be possible to make it user configurable. To change URL itself should not be too problematic, and we've seen in the past that there were no API changes always related to that. But, as I've mentioned, there could be an expected received string setting, where user can modify how it is parsed (where in received string there are particular data we look for, and which parts we don't care much about...). Something like in fail2ban log parsing in custom filters or so.

It is really problematic from my point of view if we rely on 3rd party data (where we should expect more or less frequent changes) and its usage / interpretation is hardcoded, in particular for older (non-maintained anymore) releases.

Why do you think it is soo problematic if there is a file with these settings (data) somewhere, and 99% of the users will not toch it if they don't want to? The majority of software has such files (in /etc folder), which I've never modified unless I really needed to.
